File: lib/java/projecteuler/Prime.java
package projecteuler;

import java.util.Arrays;
import static projecteuler.Factor.isFactor;

public class Prime {
  public static boolean isPrime(int n) {
    if (n <= 1) { return false; }
    for (int i = 2; i * i <= n; i++) {
      if (isFactor(i, n)) {
        return false;
      }
    }
    return true;
  }

  public static boolean[] sieve(int n) {
    boolean[] sieve = new boolean[n+1];
    Arrays.fill(sieve, true);
    sieve[0] = sieve[1] = false;
    for(int i = 2; i<=n; i++)
    {
      if(sieve[i])
      {
        for(long j = (long)i*i; j <= n; j += i) {
          sieve[(int)j] = false;
        }
      }
    }
    return sieve;
  }
}

File: lib/c++/prime.cc
#include <stdbool.h>
#include <vector>

#include "factor.h"

/*
 * Tests whether a given long is prime.
 * Works in O(sqrt(n)) time.
 * Does not work on numbers smaller than 2
 */
bool is_prime (int n) {
  if (n == 1){ return false; }
  for (int i = 2; i*i <= n; ++i) {
    if (is_factor(i, n)) {
      return false;
    }
  }
  return true;
}

std::vector<bool> sieve(int n) {
  std::vector<bool> is_prime(n + 1, true);
  is_prime[0] = false;
  is_prime[1] = false;
  for (int i = 2; i*i < n; ++i) {
    if (is_prime[i]) {
      for (int j = i+i; j < n; j += i){
        is_prime[j] = false;
      }
    }
  }       
  return is_prime;
}
